data-runtime/src/main/kotlin/io/micronaut/data/runtime/criteria/KCriteriaBuilderExt.kt

```diff
@@ -53,6 +72,9 @@ fun <E, I> From<*, E>.joinOne(prop: KProperty1<out E, I>, joinType: JoinType? =
 @Experimental
 inline fun <reified E> where(noinline dsl: Where<E>.() -> Unit) = WherePredicate(dsl)
 
+@Experimental
+inline fun <reified E> query(noinline dsl: SelectQuery<E, *>.() -> Unit) = QueryPredicate(dsl)
+
 @Experimental
 inline fun <reified E, reified R> query(noinline dsl: SelectQuery<E, R>.() -> Unit) = QueryBuilder(dsl, E::class.java, R::class.java)
```

Comment:
How is the compiler supposed to choose which method is correct?

Reply:
Depends on whether you pass two parameters or one; for example,

`query<Entity> {...}` vs. `query<Entity, ReturnType>`

We can also give the method another name, but I think same name is also fine. Basically, when passing only one parameter, we are expecting the `ReturnType` to be exactly the same as the `Entity`

Reply:
I think the correct way would be to add the missing methods to the repository interface.

Reply:
I think what you are saying is also definitely needed. But effectively, I think we should have an API that:
- Doesn't require specifying a Return Type when it's the same as the entity
- Implements a major Interface (i.e. QuerySpecification) that is allegedly used to build dynamic queries

Reply:
Can you please push a second branch with those changes that don't work? I will check

Reply:
Here you go https://github.com/micronaut-projects/micronaut-data/pull/2730

There is a spec over there that I have been testing with

Reply:
@dstepanov do you have any other concerns regarding this PR? I think the changes proposed are pretty solid and provide enhanced functionality for developers to:

- Use `query` without needing to specify a return type
- Implements the `QuerySpecification` for developers who need access to the `CriteriaQuery` instance without being able to modify its return type (`select(Nothing)`)
- More granular returned Join types (`ListJoin` and `SetJoin`)

As for the change you are proposing, I still do think it is necessary for developers who want to implement the `CoroutineJpaSpecificationExceutor` and need to return query projections (specify a return type other than the entity).

If you don't have any more concerns and agree with my statements above, are we good with merging this PR? Thanks a lot.

Reply:
Looks fine, it would be nice to add missing methods to the coroutine repository.

Reply:
It would be amazing! That's what I tried to (partially) do here: https://github.com/micronaut-projects/micronaut-data/pull/2730, but as I told you, it is unsuccessful. I think some configs need to be changed there.